Snippets/Snippets_6/UpgradeGuides/5to6/CustomErrorHandlingBehaviour.cs

```diff
@@ -0,0 +1,59 @@
+namespace Snippets6.UpgradeGuides._5to6
+{
+    using System;
+    using System.Threading.Tasks;
+    using NServiceBus;
+    using NServiceBus.Pipeline;
+
+    #region 5to6-customerrorhandlingbehaviour
+    class CustomErrorHandlingBehaviour : Behavior<ITransportReceiveContext>
+    {
+        public override async Task Invoke(ITransportReceiveContext context, Func<Task> next)
+        {
+            try
+            {
+                await next().ConfigureAwait(false);
+            }
+            catch (MessageDeserializationException)
+            {
+                Console.WriteLine("CustomFaultManager - MessageDeserializationException");
+            }
+            catch (Exception)
+            {
+                Console.WriteLine("CustomFaultManager - ProcessingAlwaysFailsForMessage");
+                // do not rethrow the message if you want to mark the message as processed.
+                // rethrow to move the message to the error queue.
+                throw;
+                // if you want to rollback the receive operation instead of mark as processed:
+                //context.AbortReceiveOperation();
+            }
+        }
+    }
+    #endregion
+
+    #region 5to6-newmessageprocessingpipelinestep
+    class NewMessageProcessingPipelineStep : RegisterStep
+    {
+        public NewMessageProcessingPipelineStep()
+            : base("CustomErrorHandlingBehaviour", typeof(CustomErrorHandlingBehaviour), "Adds custom error behavior to pipeline")
+        {
+            InsertAfter("MoveFaultsToErrorQueue");
```

Comment:
we were mainly talking about the idea of removing InsertBefore & After and how we currently need to that to emulate `IManageMessageFailures`. We both agreed that the current workaround isn't pretty and we need to find a better extensibility mechanisms. For that we need to have a better understanding on what you actually want to achieve with `IManageMessageFailures`. We don't want to reimplement IManageMessageFailures capabilities 1:1 (since that interface also seems weird) but we'd like to gather some usage scenarios where you want to modify the existing "failed message handling mechanisms" and come up with a better API.